# CRISPResso2 Workflow

Analyzing editing outcomes with CRISPResso2 on SCG
Last updated on: 2020-03-23

CRISPResso2 is a bioinformatics tool for analyzing and interpreting CRISPR genome editing experiments developed by the Pinello lab. Briefly, CRISPResso2:

  - aligns sequencing reads to a reference sequence
  - quantifies insertions, mutations and deletions to determine whether a read is modified or unmodified by genome editing
  - summarizes editing results in intuitive plots and datasets

This document is intended as a quick start guide for running CRISPResso2 using the Cong lab workflow on the [SCG cluster](https://login.scg.stanford.edu/). A companion training directory is located at `/labs/congle/training/crispresso` containing data from a knockout experiment. We will only consider a subset of the data from two SpCas9 guides targeted against AGO2. See the [Benchling entry](https://benchling.com/s/etr-VgsWIHxChinjyWovzQm7) and the included samplesheet for experimental details.

For more detailed documentation, refer to the [CRISPResso2 GitHub repository](https://github.com/pinellolab/CRISPResso2).

## Environment Setup

CRISPResso2 can be installed using the [conda](http://conda.pydata.org/docs/intro.html) package manager [Bioconda](https://bioconda.github.io/). On the cluster, first load the anaconda or miniconda module and then create a new conda environment.

```bash
module load anaconda
conda create -n csp -c bioconda crispresso2 python=2.7
```

Alternatively, a CRISPResso2 environment can be created from an exported environment file `crispresso2.yml` located in the training folder.

```bash
module load anaconda
conda env create -f crispresso2.yml
```

Verify that CRISPResso2 is installed correctly by activating the environment and running the help command.

```bash
source activate csp
CRISPResso --help
```

## Demultiplex Reads

NGS samples are often uniquely tagged with indices and pooled during sequencing. Primary sequencing data is stored as .bcl files containing the called base and confidence of the call (as a quality score). Prior to analysis, raw .bcl files must be demultiplexed to retrieve data for individual samples and converted to the text-based FASTQ sequence data format.

To do so, update sample metadata -- including `Sample_ID`, `Sample_Name`, `index`, and `index2` -- in `samples_ngs.csv` and submit the `bcl2fastq.sh` job script.

The relevant lines in the script for demultiplexing are:

```bash
WORKDIR="/labs/congle/training/crispresso"
bcl2fastq \
    --runfolder-dir "${WORKDIR}/runfolder" \
    --output-dir    "${WORKDIR}/fastq" \
    --sample-sheet  "${WORKDIR}/samples_ngs.csv" \
    --no-lane-splitting
```

This runs Illumina’s bcl2fastq tool and specifies `--runfolder-dir`containing the raw sequencing data, `--output-dir` the location to store demultiplexed FASTQ files, and the input `--sample-sheet`.

## Preprocess Reads

Demultiplexed sequencing reads need to be trimmed to remove adapter sequences and low quality bases. Additionally, paired-end reads should be merged into a single consensus read.
First, create a sample metadata file to be used as input to the trimmerge script by running `generate_trimmergefile.py` and specifying the master samplesheet.

```bash
python3 generate_trimmergefile.py [-h] -s SAMPLESHEET
                                  [--fastq_dir FASTQ_DIR]
                                  [--fastq_subdir FASTQ_SUBDIR]
                                  [--outfile OUTFILE]
```

Note that the script generates sample file paths based on information contained in the samplesheet and does not check that the paths are valid. Therefore, depending on how exactly samples were specified, the output file generated by the script may be incorrect, especially when complex pooling and sample specification layouts are used. In that case, either tweak the script to your needs or manually generate a three-column comma-separated file containing:

  - Sample name
  - Path to demultiplexed sample read1 FASTQ file
  - Path to demultiplexed sample read2 FASTQ file

Next, submit the `trimmerge.sh` script to perform read trimming and merging.
The script generates one job per file using [SLURM job arrays](https://slurm.schedmd.com/job_array.html). Therefore, the script must be edited to specify the number of jobs to run (equivalent to the number of samples). For the training dataset, we have two sample files so the array specification appears as:

```bash
#SBATCH --array=1-2
```

The lines relevant for read trimming with [Cutadapt](https://cutadapt.readthedocs.io/en/stable/) are:

```bash
cutadapt \
    -g CCATCTCATCCCTGCGTGTCTCC \
    -a CATCACCGACTGCCCATAGAGAGG \
    -G CCTCTCTATGGGCAGTCGGTGATG \
    -A GGAGACACGCAGGGATGAGATGG \
    -o "${SAMPLE}.R1.trimmed.fastq.gz" \
    -p "${SAMPLE}.R2.trimmed.fastq.gz" \
    --discard-untrimmed \
    --pair-filter=both \
    -j 0 \
       ${READ1} \
       ${READ2}
```

The lines relevant for paired-end read merging with [Flash2](https://github.com/dstreett/FLASH2) are:

```bash
/labs/congle/Software/FLASH2/flash2 \
    -M 300 \
    --allow-outies \
    --compress \
    -o ${SAMPLE} \
       ${SAMPLE}.R1.trimmed.fastq.gz \
       ${SAMPLE}.R2.trimmed.fastq.gz
```

Refer to the tools’ documentation pages for more details.

The trimmed, merged FASTQ outputs are located at `trimmerge/{SAMPLE}.extendedFrags.fastq.gz`.

## Run CRISPResso2

Now that we have demultiplexed, trimmed, and merged our sample reads, we can process them with CRISPResso2 to determine editing outcomes.

Generally, we use the CRISPRessoBatch running mode to analyze multiple files in parallel with a single job and generate summary reports across all samples. CRISPRessoBatch requires a file to specify additional parameters that will be applied to each sample when processed. As before, we can run a simple Python script to generate this batchfile:

```bash
python3 generate_batchfile.py [-h] -s SAMPLESHEET [--outfile OUTFILE]
```

Note that the currently specified fields in this file are:
  - sample name
  - path to trimmed, merged FASTQ file
  - name of the amplicon
  - sequence of the amplicon
  - sequence of the gRNA
  - cleavage offset of the nuclease

These fields may need to be changed depending on the type and parameters of the specific experiment to be analyzed. (E.g. An additional expected_hdr_amplicon_seq parameter needs to be specified for editing experiments with repair templates.)

Finally, we can submit the `crispresso_batch.sh` job script. The relevant CRISPRessoBatch lines are:

```bash
CRISPRessoBatch \
    --n_processes 16 \
    --batch_settings $BATCHFILE \
    --batch_output_folder . \
    --amplicon_min_alignment_score 60\
    --quantification_window_size 1 \
    --plot_window_size 20 \
    --skip_failed
```

CRISPRessoBatch outputs are located in the `CRISPRessoBatch_on_samples_batch` directory.
